Aggregation with implicit grouping keys are deprecated (#1293)
neo4j/neo4j-documentation: cypher/cypher-docs/src/docs/dev/deprecations-additions-and-compatibility.asciidoc

| Details

a|
label:syntax[]
label:added[]
[source, cypher, role="noheader"]
----
MATCH (n) RETURN n.propertyName_1, n.propertyName_2 + count(*)
----
a|

> Aggregation with implicit grouping keys are deprecated. In an expression that contains aggregation, the leaves of expressions must be one of: +

Reviewer: "Implicit grouping keys" has a bit of a double meaning.

a) grouping key can be considered implicit if they are not given with an explicit grouping keys -- these are not deprecated
b) grouping key can be considered implicit if they are inferred from subexpression of aggregation expressions -- these are the once we are deprecating

Variant b) is obviously more implicit than the variant a).

I do not have immediate idea how to improve the text in this regard. The best I can come up with is:

```suggestion
Implied grouping keys are deprecated. Only expressions that do _not_ contain aggregations are grouping keys. In expressions that contain aggregations, the leaves of expressions must be one of: +
```

> Implied grouping keys are deprecated. Only expressions that do _not_ contain aggregations are grouping keys. In expressions that contain aggregations, the leaves of expressions must be one of: +

Reviewer 2: ```suggestion
Implied grouping keys are deprecated. 
Note that only expressions that do _not_ contain aggregations are grouping keys. 
In expressions that contain aggregations, the leaves of expressions must be one of: +
```

- An aggregation +
- A literal +
- A Parameter +

> - A variable - ONLY IF that variable is also one of: +

Reviewer 2: ```suggestion
- A variable - *ONLY IF* that variable is also one of: +
```

>     1) a projection expression on its own (e.g. the `n` in `RETURN n AS myNode, n.value + count(*)`) OR +

Reviewer: ```suggestion
    1) A projection expression on its own (e.g. the `n` in `RETURN n AS myNode, n.value + count(*)`) OR +
```

>     1) A projection expression on its own (e.g. the `n` in `RETURN n AS myNode, n.value + count(*)`) OR +

Reviewer 2: ```suggestion
    1) A projection expression on its own (e.g. the `n` in `RETURN n AS myNode, n.value + count(\*)`) OR +
```

Reviewer 2: The asterisk needs to be escaped or else it will make the text bold.

>     2) A local variable in the expression (e.g the `x` in `RETURN n, n.prop + size([ x IN range(1, 10) | x ])` +

Reviewer 2: ```suggestion
    2) A local variable in the expression (e.g the `x` in `RETURN n, n.prop + size([ x IN range(1, 10) \| x ])` +
```

Reviewer 2: The pipe in the expression is what made everything after it appear in the next column, so it needs to be escaped.

Author: aha... 🙄 😓

> - Property access - ONLY IF that property access is also a projection expression on its own (e.g. the `n.prop` in `RETURN n.prop, n.prop + count(*)`) +

Reviewer 2: ```suggestion
- Property access - *ONLY IF* that property access is also a projection expression on its own (e.g. the `n.prop` in `RETURN n.prop, n.prop + count(*)`) +
```

Reviewer 2: As above

> - Map access - ONLY IF that map access is also a projection expression on its own (e.g. the `map.prop` in `WITH {prop: 2} AS map RETURN map.prop, map.prop + count(*)`)

Author: Oh, I now see that it seems like this text is too long and that it's overflowing to the next features text 😨  Do you know how to fix that @AlexicaWright?

Reviewer 2: Yep! See above ⬆️

Reviewer 2: ```suggestion
- Map access - *ONLY IF* that map access is also a projection expression on its own (e.g. the `map.prop` in `WITH {prop: 2} AS map RETURN map.prop, map.prop + count(*)`)
```
